src/function/algebra/derivative.js

```diff
@@ -143,60 +152,51 @@ export const createDerivative = /* #__PURE__ */ factory(name, dependencies, ({
-   * constNodes. Classification is done as follows:
+   * Checks if a node is constants (e.g. 2 + 2).
+   * Accepts (usually memoized) version of self as the first parameter for recursive calls.
+   * Classification is done as follows:
    *
    *   1. ConstantNodes are constants.
    *   2. If there exists a SymbolNode, of which we are differentiating over,
    *      in the subtree it is not constant.
    *
-   * @param  {Object} constNodes  Holds the nodes that are constant
+   * @param  {function} isConst  Function that tells whether sub-expression is a constant
    * @param  {ConstantNode | SymbolNode | ParenthesisNode | FunctionNode | OperatorNode} node
    * @param  {string} varName     Variable that we are differentiating
    * @return {boolean}  if node is constant
    */
-  // TODO: can we rewrite constTag into a pure function?
-  const constTag = typed('constTag', {
-    'Object, ConstantNode, string': function (constNodes, node) {
-      constNodes[node] = true
+  const _isConst = typed('_isConst', {
+    'function, ConstantNode, string': function () {
       return true
     },
 
-    'Object, SymbolNode, string': function (constNodes, node, varName) {
+    'function, SymbolNode, string': function (isConst, node, varName) {
       // Treat other variables like constants. For reasoning, see:
       //   https://en.wikipedia.org/wiki/Partial_derivative
-      if (node.name !== varName) {
-        constNodes[node] = true
-        return true
-      }
-      return false
+      return node.name !== varName
     },
 
-    'Object, ParenthesisNode, string': function (constNodes, node, varName) {
-      return constTag(constNodes, node.content, varName)
+    'function, ParenthesisNode, string': function (isConst, node, varName) {
+      return isConst(node.content, varName)
     },
 
-    'Object, FunctionAssignmentNode, string': function (constNodes, node, varName) {
+    'function, FunctionAssignmentNode, string': function (isConst, node, varName) {
       if (!node.params.includes(varName)) {
-        constNodes[node] = true
         return true
       }
-      return constTag(constNodes, node.expr, varName)
+      return isConst(node.expr, varName)
     },
 
-    'Object, FunctionNode | OperatorNode, string': function (constNodes, node, varName) {
+    'function, FunctionNode | OperatorNode, string': function (isConst, node, varName) {
       if (node.args.length > 0) {
-        let isConst = constTag(constNodes, node.args[0], varName)
-        for (let i = 1; i < node.args.length; ++i) {
-          isConst = constTag(constNodes, node.args[i], varName) && isConst
+        let allConst = true
+        for (let i = 0; i < node.args.length && allConst; ++i) {
+          allConst = isConst(node.args[i], varName) && allConst
         }
 
-        if (isConst) {
-          constNodes[node] = true
-          return true
-        }
+        return allConst
       }
+      // TODO: add a comment explaining why false (and why is this reachable)
```

Comment:
Can you either write the the comment or cleanup the TODO? (the comment is not neccesary from my point of view, but can be helpful indeed).

Reply:
Removed it. TBH I'm personally not sure what's going on here -- it looks like a function call without arguments, which I'd expect to be a constant (unless `f` is an alias. can expressions do that?)
But since I've successfully sort of tricked you into looking here I'm happy to put this behind :)